README.md
# egg-utils

[![NPM version][npm-image]][npm-url]
[![npm download][download-image]][download-url]
[![Build Status](https://www.travis-ci.org/deepexi/egg-utils.svg?branch=master)](https://www.travis-ci.org/deepexi/egg-utils)
[![codecov](https://codecov.io/gh/deepexi/egg-utils/branch/master/graph/badge.svg)](https://codecov.io/gh/deepexi/egg-utils)

[npm-image]: https://img.shields.io/npm/v/@taccisum/egg-utils.svg
[npm-url]: https://www.npmjs.com/package/@taccisum/egg-utils
[download-image]: https://img.shields.io/npm/dm/@taccisum/egg-utils.svg
[download-url]: https://www.npmjs.com/package/@taccisum/egg-utils

[CHANGELOG](./CHANGELOG.md)

## 依赖的插件

- circuit-fuses v4.x
- request v2.x

## 如何使用

添加依赖

```bash
$ npm i @taccisum/egg-utils --save
```

开启插件

```js
// config/plugin.js
exports.utils = {
  enable: true,
  package: 'egg-utils',
};
```

### http utils

简单请求：

```js
const resp = await ctx.helper.http.requestGet('http://www.baidu.com')
```
### sse utils

注意：node长连接默认2分钟会超时，如服务端连接发送事件的间隔需超过2分钟，需服务端自行处理（设置timeout时间 or 在2分钟的间隔内发送无效事件）

#### 使用示例

##### server端
###### 单次发送消息（如状态需实时返回给前端）
```js
/* router */
app.get('logs', `/${APP_CONTEXT}/api/v2/logs`, app.controller.v1.getLog);
/* router */
const SSEUtils = ctx.helper.sse;
/* controller */
async getLog () {
  const ctx = this.ctx;
  ctx.body = SSEUtils.send({
    setResHeader: resHeaders => {
      ctx.set(resHeaders);
    },
    sendType: 'once',
    onceMsg: 'hello world',
  });
}
```
###### 其他情况发送消息（非单次，或者其他任意情况）
```js
/* router */
app.get('logs', `/${APP_CONTEXT}/api/v2/logs`, app.controller.v1.getLog);
/* router */
const SSEUtils = ctx.helper.sse;
/* controller */
async getLog () {
  const ctx = this.ctx;
  const msgs = ['hello' 'world'];
  let interval = null
  ctx.body = SSEUtils.send({
    setResHeader: resHeaders => {
      ctx.writeHead(resHeaders);
    },
    sendType: 'other',
    sender: send => {
      let index = 0;
      interval = setInterval(() => {
        const msg = index === msgs.length ? 'sseEnd' : msgs[index];
        send(msg);

        if (msg === 'sseEnd') {
          // 清除事件
          clearInterval(interval);
        }
        index++;
      }, 1000); //每秒发送一次消息
    },
  })
}
```

#### API
##### Methods

|   方法名   | 说明 | 参数 |
| :--: | :--: | ---- |
| send | 与客户端建立长连接,返回值是 stream对象 | optsions选项<br />options.setResHeader 设置请求头function，required<br />options.sendType once单次 repeat重复 other其他，默认发送一次<br />options.sender 消息发送者，处理什么时候发送消息和结束发送消息，参数send func，结束需要发送'sseEnd'消息，非一次使用<br />options.onceMsg 单次发送消息主体，默认是''<br />options.retry 长连接发送错误时，重试频率，毫秒, 默认10s<br /> options.msgReplace 无法处理消息带换行符的情况，提供替换的正则，默认为''<br /> options.finishCb stream结束时钩子函数 |

### crypto utils

#### 使用示例

##### md5加密
```js
const CryptoUtils = ctx.helper.crypto;
const hashData = CryptoUtils.md5('data', 'salt');
```
##### aes加密
```js
const CryptoUtils = ctx.helper.crypto;
const data = 'data';
const secretKey = 'secretKeyDataKey';   // 秘钥必须16位
const iv = 'secretKeyDataKey';          // 初始向量 initial vector 必须16位
CryptoUtils.aesEncrypt(data, secretKey, iv);
```

##### aes解密
```js
const CryptoUtils = ctx.helper.crypto;
const data = 'data';
const secretKey = 'secretKeyDataKey';   // 加密时的秘钥 必须16位
const iv = 'secretKeyDataKey';          // 加密时的初始向量 initial vector 必须16位
CryptoUtils.aesDecrypt(data, secretKey, iv);
```

##### aes加解密对象
```js
const CryptoUtils = ctx.helper.crypto;
CryptoUtils.getCipheriv(algorithm, secretKey, iv);
CryptoUtils.getDecipheriv(algorithm, secretKey, iv);
```

## License

[MIT](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.3.0](https://github.com/deepexi/egg-utils/compare/v0.2.1...v0.3.0) (2019-08-15)


### Bug Fixes

* err test cases ([07b5b04](https://github.com/deepexi/egg-utils/commit/07b5b04))


### Features

* provide crypto utils ([#4](https://github.com/deepexi/egg-utils/issues/4)) ([338346e](https://github.com/deepexi/egg-utils/commit/338346e))



### [0.2.1](https://github.com/deepexi/egg-utils/compare/v0.2.0...v0.2.1) (2019-08-08)



## [0.2.0](https://github.com/eggjs/egg-utils/compare/v0.1.1...v0.2.0) (2019-08-08)


### Features

* add optional finish callback fn to sse_utils.send ([88e87b6](https://github.com/eggjs/egg-utils/commit/88e87b6))



### [0.1.1](https://github.com/eggjs/egg-utils/compare/v0.1.0...v0.1.1) (2019-07-18)


### Bug Fixes

* sse utils bug ([43058bc](https://github.com/eggjs/egg-utils/commit/43058bc))


test/utils.test.js
'use strict';

const assert = require('assert');
const mock = require('egg-mock');
const HttpUtils = require('../app/extend/utils/http_utils');

describe('test/utils.test.js', () => {
  describe('integrate with eggjs', () => {
    let app;
    before(() => {
      app = mock.app({
        baseDir: 'apps/utils-test',
      });
      return app.ready();
    });

    after(() => app.close());
    afterEach(mock.restore);

    it('should load utils /', () => {
      const ctx = app.mockContext();
      assert(ctx.helper.http, 'http utils should not be undefined!');
    });

    it('should get /', async () => {
      const ctx = app.mockContext();
      const r = await ctx.helper.http.requestGet('http://www.baidu.com');
      assert(r, 'response should not be empty');
    });
  });

  describe('HttpUtils', () => {
    describe('runCircuitVerb()', () => {
      const clog = msg => {
        console.log(msg);
      };

      const getUtils = () => {
        return new HttpUtils({
          debug: clog,
          info: clog,
          warn: clog,
          error: clog,
        });
      };

      describe('_getVerbFn()', () => {
        let utils;

        before(() => {
          utils = getUtils();
        });

        it('should get on default', () => {
          assert(utils._getVerbFn() === require('request').get);
        });

        it('should right when verb incollected', () => {
          assert(utils._getVerbFn('gEt') === require('request').get);
        });

        it('should right by verb', () => {
          assert(utils._getVerbFn('get') === require('request').get);
          assert(utils._getVerbFn('post') === require('request').post);
          assert(utils._getVerbFn('put') === require('request').put);
          assert(utils._getVerbFn('delete') === require('request').delete);
        });

        it('should error when unsupport verb', () => {
          try {
            utils._getVerbFn('unknown_verb');
            assert.fail();
          } catch (e) {
            // assert pass if catch error
          }
        });
      });
    });
  });
});


test/sse_utils.test.js
'use strict';

const assert = require('assert');
const SSEUtils = require('../app/extend/utils/sse_utils');

/* eslint-disable no-new */
const EventSource = require('eventsource');
const http = require('http');

let _port = 20000;
const servers = [];
process.on('exit', function() {
  if (servers.length > 0) {
    console.error(
      "************ Didn't kill all servers - there is still %d running.",
      servers.length
    );
  }
});

function createServer(callback) {
  const server = http.createServer();
  configureServer(server, 'http', _port++, callback);
}

function configureServer(server, protocol, port, callback) {
  const responses = [];

  const oldClose = server.close;
  server.close = function(closeCb) {
    responses.forEach(function(res) {
      res.end();
    });
    oldClose.call(this, function() {
      servers.splice(servers.indexOf(server), 1);
      closeCb && closeCb();
    });
  };

  server.on('request', function(req, res) {
    responses.push(res);
  });

  server.url = protocol + '://localhost:' + port;
  server.listen(port, function onOpen(err) {
    servers.push(server);
    callback(err, server);
  });
}

function writeEventsOnce(msg, finishCb) {
  return (req, res) => {
    const stream = SSEUtils.send({
      setResHeader: resHeaders => {
        res.writeHead(200, resHeaders);
      },
      sendType: 'once',
      onceMsg: msg,
      finishCb,
    });

    res.body = stream.pipe(res);
  };
}

function writeEventsArr(msgs) {
  return (req, res) => {
    const stream = SSEUtils.send({
      setResHeader: resHeaders => {
        res.writeHead(200, resHeaders);
      },
      sendType: 'other',
      sender: send => {
        msgs.forEach((msg, index) => {
          send(msg);
          if (index === msgs.length - 1) {
            send('sseEnd');
          }
        });
      },
    });

    res.body = stream.pipe(res);
  };
}
function writeEventsInterval(msgs, times = 1000, finishCb) {
  let interval = null;
  return (req, res) => {
    const stream = SSEUtils.send({
      setResHeader: resHeaders => {
        res.writeHead(200, resHeaders);
      },
      sendType: 'other',
      sender: send => {
        let index = 0;
        interval = setInterval(() => {
          const msg = index === msgs.length ? 'sseEnd' : msgs[index];
          send(msg);

          if (msg === 'sseEnd') {
            // 清除事件
            clearInterval(interval);
          }
          index++;
        }, times);
      },
      finishCb,
    });

    res.body = stream.pipe(res);
  };
}

describe('send msg', function() {
  it('should right when send msg once', function(done) {
    createServer(function(err, server) {
      if (err) return done(err);

      server.on('request', writeEventsOnce('hello world'));
      const es = new EventSource(server.url);

      es.onmessage = function(m) {
        assert.equal('hello world', m.data);
        server.close(done);
      };
    });
  });

  it('should right when send msg arr', function(done) {
    createServer(function(err, server) {
      if (err) return done(err);
      const msgs = [ 'hello', 'world' ];
      server.on('request', writeEventsArr(msgs));
      const es = new EventSource(server.url);
      let times = 0;
      es.onmessage = function(m) {
        times++;
        // 预计第二条信息获取为world
        if (times === 2) {
          assert.equal('world', m.data);
        }

        if (msgs.length === times) {
          server.close(done);
        }
      };
    });
  });

  it('should right when send msg interval', function(done) {
    createServer(function(err, server) {
      if (err) return done(err);
      const msgs = [ 'hello', 'world' ];
      server.on('request', writeEventsInterval(msgs, 1000));
      const es = new EventSource(server.url);
      let times = 0;
      es.onmessage = function(m) {
        times++;
        // 预计第二条信息获取为world
        if (times === 2) {
          assert.equal('world', m.data);
        }

        if (msgs.length === times) {
          server.close(done);
        }
      };
    });
  });
});

describe('sse finishCb', function() {
  it('should right when send msg once finishCb', function(done) {
    createServer(function(err, server) {
      if (err) return done(err);

      let finishFlag = false;
      server.on('request', writeEventsOnce('hello world', () => {
        finishFlag = true; // stream关闭标签
        assert.equal(true, finishFlag);
        done();
      }));

      const es = new EventSource(server.url);
      es.onmessage = () => {};

      // 主动关闭http request
      setTimeout(() => {
        server.close();
      }, 100);
    });
  });

  it('should right when send msg interval finishCb', function(done) {
    createServer(function(err, server) {
      if (err) return done(err);
      const msgs = [ 'finish01', 'finish02', 'finish03' ];
      let finishFlag = false;
      server.on('request', writeEventsInterval(msgs, 100, () => {
        finishFlag = true;
        assert.equal(true, finishFlag);
        done();
      }));

      const es = new EventSource(server.url);
      es.onmessage = () => {};

      // 主动关闭http request
      setTimeout(() => {
        server.close();
      }, 200);
    });
  });
});

describe('sse error', function() {
  it('should error when send msg do not provide setResHeader option', function(done) {
    createServer(function(err, server) {
      if (err) return done(err);

      let errorFlag = false;
      server.on('request', (req, res) => {
        try {
          const stream = SSEUtils.send({
            sendType: 'other',
          });
          res.body = stream.pipe(res);
        } catch (e) {
          errorFlag = true;
          assert.equal(true, errorFlag);
          done();
        }
      });
      const es = new EventSource(server.url);
      es.onmessage = () => {};
    });
  });

  it('should error when send other type msg do not provide sender option', function(done) {
    createServer(function(err, server) {
      if (err) return done(err);

      let errorFlag = false;
      server.on('request', (req, res) => {
        try {
          const stream = SSEUtils.send({
            setResHeader: resHeaders => {
              res.writeHead(200, resHeaders);
            },
            sendType: 'other',
          });
          res.body = stream.pipe(res);
        } catch (e) {
          errorFlag = true;
          assert.equal(true, errorFlag);
          done();
        }
      });
      const es = new EventSource(server.url);
      es.onmessage = () => {};
    });
  });
});



test/crypto_utils.test.js
'use strict';

const DigestUtils = require('../app/extend/utils/crypto_utils');
const assert = require('assert');


describe('#crypto_utils.js', () => {

  it('should be return string without salt value', function() {
    const md5 = DigestUtils.md5('data');
    assert.equal(md5.length, 32);
  });

  it('should be return string with salt value', function() {
    const md5 = DigestUtils.md5('data', 'dada');
    assert.equal(md5.length, 32);
  });

  it('should be throw error', function() {
    assert.throws(() => {
      DigestUtils.md5({ data: 'data' }, 'dada');
    });
  });

  it('should be throw error salt is must be string', function() {
    assert.throws(() => {
      DigestUtils.md5('data', { salt: 'salt' });
    });
  });


  // aes加密
  it('aes should be throw error secretKey or iv length must be 16', function() {
    assert.throws(() => {
      DigestUtils.aesEncrypt('data', 'abcd', 'abc');
    });
  });

  it('aes should be return string', function() {
    const aes = DigestUtils.aesEncrypt('data', 'abcdabcdabcdabcd', 'abcdabcdabcdabcd');
    assert.equal(aes.length, 32);
  });

  it('aes should be throw error data type is not suitable', function() {
    assert.throws(() => {
      DigestUtils.aesEncrypt({ data: 'data' }, 'abcdabcdabcdabcd', 'abcdabcdabcdabcd');
    });
  });

  // aes 解密
  it('should be return data', function() {
    const aes = DigestUtils.aesEncrypt('data', 'abcdabcdabcdabcd', 'abcdabcdabcdabcd');
    const result = DigestUtils.aesDecrypt(aes, 'abcdabcdabcdabcd', 'abcdabcdabcdabcd');
    assert.equal(result, 'data');
  });

  it('should be return null is not suitable', function() {
    const aes = DigestUtils.aesEncrypt('data', 'abcdabcdabcdabcd', 'abcdabcdabcdabcd');
    const result = DigestUtils.aesDecrypt(aes, 'abcdabcdabcdabda', 'abcdabcdabcdabcd');
    assert.equal(result, null);
  });

  it('should be throw error dataType is not suitable', function() {
    const aes = DigestUtils.aesEncrypt('data', 'abcdabcdabcdabcd', 'abcdabcdabcdabcd');
    assert.throws(() => {
      DigestUtils.aesDecrypt({ aes }, 'abcdabcdabcdabda', 'abcdabcdabcdabcd');
    });
  });

  it('should be return a cipher', function() {
    const aes = DigestUtils.getCipheriv('aes-128-cbc', 'abcdabcdabcdabcd', 'abcdabcdabcdabcd');
    let result;
    if (typeof aes.update === 'function') {
      result = true;
    }
    assert.equal(result, true);
  });

  it('should be return a decipher', function() {
    const aes = DigestUtils.getDecipheriv('aes-128-cbc', 'abcdabcdabcdabcd', 'abcdabcdabcdabcd');
    let result;
    if (typeof aes.update === 'function') {
      result = true;
    }
    assert.equal(result, true);
  });


});


app/extend/utils/http_utils.js
'use strict';

const request = require('request');
const Breaker = require('circuit-fuses').breaker;

class ResponseError extends Error {
  constructor(msg, res) {
    super(msg);
    this.response = res;
  }
}

class HttpUtils {
  constructor(logger) {
    this.logger = logger;
  }

  _getVerbFn(verb) {
    if (!verb) verb = 'GET';
    const command = request[(verb + '').trim().toLowerCase()];

    if (!command) {
      throw new Error('unsupported verb for http request');
    }
    return command;
  }

  _proxyCommand(command) {
    return new Breaker(command);
  }

  /**
   * request + 熔断器
   * @param {String} verb 动词
   * @param {String} url url
   * @param {Object} options 选项
   * @return {Promise<void>}
   */
  async runCircuitVerb(verb = 'GET', url, options) {
    const command = this._getVerbFn(verb);
    const breaker = this._proxyCommand(command);
    try {
      this.logger.debug(`外部请求url:${url},options:${JSON.stringify(options)}`);
      const res = await breaker.runCommand(url, options);
      if (res.statusCode >= 200 && res.statusCode < 300) return res.body;
      throw new ResponseError('响应错误', res);
    } catch (err) {
      // TODO:: 有关降级的策略 err.message = "CircuitBreaker Open"
      /* If the circuit is open the command is not run,
       * and an error with message "CircuitBreaker Open" is returned.
       * In this case, you can switch on the error and have a fallback technique
       */
      this.logger.warn(`${verb}:${url} ${err.message}`);
      throw err;
    }
  }

  /**
   * 发送http get 请求
   * @param {String} url url
   * @param {Object} qs 查询参数对象
   * @param {Object} options 选项
   * @return {Promise<*>} response
   */
  async requestGet(url, qs, options = { timeout: 30000, json: true }) {
    if (qs) options.qs = qs;
    return await this.runCircuitVerb('GET', url, options);
  }

  /**
     * 发送http post 请求
     * @param {String} url url
     * @param {Object} body request请求体
     * @param {Object} options 选项
     * @return {Promise<*>} response
     */
  async requestPost(url, body, options = { timeout: 150000, json: true }) {
    if (body) options.body = body;
    return await this.runCircuitVerb('POST', url, options);
  }

  /**
   * 发送http put 请求
   * @param {String} url url
   * @param {Object} body request请求体
   * @param {Object} options 选项
   * @return {Promise<*>} response
   */
  async requestPut(url, body, options = { timeout: 30000, json: true }) {
    if (body) options.body = body;
    return await this.runCircuitVerb('PUT', url, options);
  }
  /**
   * 发送http DELETE 请求
   * @param {String} url url
   * @param {Object} options 选项
   * @return {Promise<*>} response
   */
  async requestDelete(url, options = { timeout: 30000, json: true }) {
    return await this.runCircuitVerb('DELETE', url, options);
  }
}

module.exports = HttpUtils;


app/extend/utils/crypto_utils.js
/**
 * author:  zhangyongbin
 * createTime:    2019-08-12
 * updateTime:    2019-08-12
 * desc:    DigestUtils
 * version: 1.0.0
 */

'use strict';

const crypto = require('crypto');

class DigestUtils {

  /**
   * md5加密
   * @param {string} data      加密数据
   * @param {string} salt    可选对象，加盐值，salt，默认为''
   * @return {string}    加密后字符串
   */
  static md5(data, salt = '') {
    if (typeof (data) !== 'string') {
      throw new Error('data必须为字符串！');
    }
    if (typeof (salt) !== 'string') {
      throw new Error('salt必须为字符串！');
    }
    const saltData = `${data}${salt}`;
    const hash = crypto.createHash('md5');
    hash.update(saltData);
    return hash.digest('hex');
  }

  /**
   * aes加密
   * @param {string}data          加密的数据
   * @param {string}secretKey     密钥16位
   * @param {string}iv            初始向量 initial vector 16 位
   * @return {string}             加密后字符串
   */
  static aesEncrypt(data, secretKey, iv) {
    if (secretKey.length !== 16 || iv.length !== 16) {
      throw new Error('secretKey和iv长度必须为16');
    }
    const algorithm = 'aes-128-cbc';
    secretKey = Buffer.from(secretKey, 'utf8');
    iv = Buffer.from(iv, 'utf8');
    const cipher = crypto.createCipheriv(algorithm, secretKey, iv);
    let crypted = cipher.update(data, 'utf8', 'hex');
    crypted += cipher.final('hex');
    return crypted;
  }

  /**
   * aes解密
   * @param {string}data          解密的数据
   * @param {string}secretKey     密钥 16 位
   * @param {string}iv            初始向量 initial vector 16 位
   * @return {string}    解密后字符串
   */
  static aesDecrypt(data, secretKey, iv) {
    if (secretKey.length !== 16 || iv.length !== 16) {
      throw new Error('secretKey和iv长度必须为16');
    }
    const algorithm = 'aes-128-cbc';
    secretKey = Buffer.from(secretKey, 'utf8');
    iv = Buffer.from(iv, 'utf8');
    const decipher = crypto.createDecipheriv(algorithm, secretKey, iv);
    let decrypted = decipher.update(data, 'hex', 'utf8');
    try {
      decrypted += decipher.final('utf8');
    } catch (e) {
      return null;
    }
    return decrypted;
  }

  /**
   * 获取aes加密对象
   * @param {string}algorithm 算法
   * @param {string}secretKey 秘钥
   * @param {string}iv        初始向量 initial vector
   * @return {Cipher}         对象Cipher
   */
  static getCipheriv(algorithm, secretKey, iv) {
    return crypto.createCipheriv(algorithm, secretKey, iv);
  }

  /**
   * 获取aes解密对象
   * @param {string}algorithm   算法
   * @param {string}secretKey   秘钥
   * @param {string}iv          初始向量 initial vector
   * @return {Decipher}         对象Decipher
   */
  static getDecipheriv(algorithm, secretKey, iv) {
    return crypto.createDecipheriv(algorithm, secretKey, iv);
  }

}

module.exports = DigestUtils;


app/extend/utils/sse_utils.js
/**
 * author:  qiye
 * createTime:    2019-06-12
 * updateTime:    2019-08-07
 * desc:    sse utils
 * version: 0.0.7
 */

'use strict';

const PassThrough = require('stream').PassThrough;

class SSEUtils {
  /**
   * SSE http协议发送消息
   * @param {Object} options 选项
   * @param {Function} options.setResHeader 设置请求头function，必须
   * @param {String} options.sendType once单次 repeat重复 other其他，默认发送一次
   * @param {String} options.sender 消息发送者，处理什么时候发送消息和结束发送消息，参数sendMsg func
   * @param {String} options.onceMsg 单次发送消息主体，默认是''
   * @param {Number} options.retry 长连接发送错误时，重试频率，毫秒, 默认10s
   * @param {RegExp} options.msgReplace 无法处理消息带换行符的情况，提供替换的正则，默认为''
   * @param {Function} options.finishCb stream主动关闭时，回调function
   * @return {Object} stream 流对象
   */
  static send(options = {}) {
    const {
      setResHeader,
      sendType = 'once',
      sender,
      onceMsg = '',
      retry = 10000,
      msgReplace = '',
      finishCb,
    } = options;
    if (typeof setResHeader !== 'function') {
      throw new Error('请传入setResHeader！');
    }

    // 设置响应头
    setResHeader.call(this, {
      'Content-Type': 'text/event-stream',
      'Access-Control-Allow-Origin': '*',
      'Cache-Control': 'no-cache',
      Connection: 'keep-alive',
    });

    // 转换流
    const streamOptions = {
      highWaterMark: 1024 * 1024 * 1024, // 默认缓冲区大小
    };

    // 增加stream参数
    const stream = new PassThrough(Object.assign(streamOptions, options.streamOptions));

    // 替换换行符
    const replaceMsg = (msg = '') => {
      return msg.replace(/\n/g, '\r').replace(/\r/g, msgReplace);
    };

    // stream 是否可写
    const isStreamWritable = () => {
      return stream && stream.writable;
    };

    // 监听stream finish事件
    stream.on('finish', () => {
      if (typeof finishCb === 'function') {
        finishCb.call(this);
        stream.end();
      }
    });

    const endCall = () => {
      if (!isStreamWritable()) {
        return;
      }
      stream.write('event: sseEnd\n');
      stream.write('data: \n\n'); // 多发一条信息，是sseEnd事件能成功被接收
      // 关闭stream
      stream.end();
    };

    // 消息发送一次
    const sendOnce = (onceMsg = '') => {
      if (!isStreamWritable()) {
        return;
      }
      // 声明消息id合重连时间
      stream.write(`id: ${+new Date()}\n`); // 消息ID
      stream.write(`retry: ${retry || 10000}\n`); // 重连时间，默认10s
      // 首次发送消息
      stream.write(`data: ${replaceMsg(onceMsg)}`); // 消息数据
      stream.write('\n\n'); // 消息结束
      endCall.call(this);
    };

    // 其他频率发送消息，msg为sseEnd，则结束发送消息
    const sendMsg = msg => {
      if (!isStreamWritable()) {
        return;
      }
      if (msg === 'sseEnd') {
        endCall.call(this);
        return;
      }
      stream.write(`data: ${replaceMsg(msg)}`); // 消息数据
      stream.write('\n\n'); // 消息结束
    };

    switch (sendType) {
      default:
      case 'once':
        sendOnce.call(this, onceMsg);
        break;
      // case 'repeat':
      case 'other':
        // sendType other需要sender
        if (typeof sender !== 'function') {
          throw new Error('请传入func sender，消息发送者！');
        }

        stream.write(`id: ${+new Date()}\n`); // 消息ID
        stream.write(`retry: ${retry || 10000}\n`); // 重连时间，默认10s
        // 发送消息
        sender.call(this, sendMsg);
    }

    return stream;
  }
}

module.exports = SSEUtils;
